app/ts/utils/SignalEffectWithAbortOnChange.ts

```diff
@@ -0,0 +1,43 @@
+import { Signal } from '@preact/signals'
+import { OptionalSignal } from './OptionalSignal.js'
+import { useEffect } from 'preact/hooks'
+
+type UnionSignalType<T> = T extends OptionalSignal<infer U> ? U | undefined : T extends Signal<infer U> ? U extends Signal<infer Inner> ? Inner : U : never
+type SignalTypes<T extends readonly [...(Signal<unknown> | OptionalSignal<unknown>)[]]> = { [K in keyof T]: UnionSignalType<T[K]> }
+const isOptionalSignal = <T>(signal: Signal<T> | OptionalSignal<T>): signal is OptionalSignal<T> => signal instanceof OptionalSignal
```

Comment:
You can import `isOptionalSignal` from the other file instead of duplicating it in this one.
```suggestion
import { Signal } from '@preact/signals'
import { OptionalSignal, isOptionalSignal } from './OptionalSignal.js'
import { useEffect } from 'preact/hooks'

type UnionSignalType<T> = T extends OptionalSignal<infer U> ? U | undefined : T extends Signal<infer U> ? U extends Signal<infer Inner> ? Inner : U : never
type SignalTypes<T extends readonly [...(Signal<unknown> | OptionalSignal<unknown>)[]]> = { [K in keyof T]: UnionSignalType<T[K]> }
```